integration/release_upgrade_linux_test.go

```diff
@@ -0,0 +1,364 @@
+					}
+
+				}
+
+			case secondSB:
+				assert.Equal(t, criruntime.PodSandboxState_SANDBOX_NOTREADY, pod.State)
+			default:
+				t.Errorf("unexpected pod %s", pod.Id)
+			}
+		}
+	}
+}
+
+// cleanupPods deletes all the pods based on the cri.RuntimeService connection.
+func cleanupPods(t *testing.T, criRuntimeService cri.RuntimeService) {
+	pods, err := criRuntimeService.ListPodSandbox(nil)
+	require.NoError(t, err)
+
+	for _, pod := range pods {
+		assert.NoError(t, criRuntimeService.StopPodSandbox(pod.Id))
+		assert.NoError(t, criRuntimeService.RemovePodSandbox(pod.Id))
+	}
+}
+
+// currentReleaseCtrdDefaultConfig generates empty(default) config for current release.
+func currentReleaseCtrdDefaultConfig(t *testing.T, targetDir string) {
+	fileName := filepath.Join(targetDir, "config.toml")
+	err := os.WriteFile(fileName, []byte(""), 0600)
+	require.NoError(t, err, "failed to create config for current release")
+}
+
+// previousReleaseCtrdConfig generates containerd config with previous release
+// shim binary.
+func previousReleaseCtrdConfig(t *testing.T, previousReleaseBinDir, targetDir string) {
+	// TODO(fuweid):
+	//
+	// We should choose correct config version based on previous release.
+	// Currently, we're focusing on v1.x -> v2.0 so we use version = 2 here.
+	rawCfg := fmt.Sprintf(`
+version = 2
+
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
+  runtime_type = "%s/bin/containerd-shim-runc-v2"
+`,
+		previousReleaseBinDir)
+
+	fileName := filepath.Join(targetDir, "config.toml")
+	err := os.WriteFile(fileName, []byte(rawCfg), 0600)
+	require.NoError(t, err, "failed to create config for previous release")
+}
+
+// criRuntimeService returns cri.RuntimeService based on the grpc address.
+func (p *ctrdProc) criRuntimeService(t *testing.T) cri.RuntimeService {
+	service, err := remote.NewRuntimeService(p.grpcAddress(), 1*time.Minute)
+	require.NoError(t, err)
+	return service
+}
+
+// criImageService returns cri.ImageManagerService based on the grpc address.
+func (p *ctrdProc) criImageService(t *testing.T) cri.ImageManagerService {
+	service, err := remote.NewImageService(p.grpcAddress(), 1*time.Minute)
+	require.NoError(t, err)
+	return service
+}
+
+// newCtrdProc is to start containerd process.
+func newCtrdProc(t *testing.T, ctrdBin string, ctrdWorkDir string) *ctrdProc {
+	p := &ctrdProc{workDir: ctrdWorkDir}
+
+	var args []string
+	args = append(args, "--root", p.rootPath())
+	args = append(args, "--state", p.statePath())
+	args = append(args, "--address", p.grpcAddress())
+	args = append(args, "--config", p.configPath())
+	args = append(args, "--log-level", "debug")
+
+	f, err := os.OpenFile(p.logPath(), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0600)
+	require.NoError(t, err, "open log file %s", p.logPath())
+	t.Cleanup(func() { f.Close() })
+
+	cmd := exec.Command(ctrdBin, args...)
+	cmd.Stdout = f
+	cmd.Stderr = f
+	cmd.SysProcAttr = &syscall.SysProcAttr{Pdeathsig: syscall.SIGKILL}
```

Comment:
It looks like there's some subtlety around pdeathsig in Go: https://github.com/golang/go/issues/27505#issuecomment-713706104.  It might be worth implementing something like this and using a channel for wait inside a locked thread.

Reply:
Updated. PTAL